### PR TITLE
Bandit tutorials

### PR DESCRIPTION
**Pull request by Darshan-ko**
@sampreet-arthi @threewisemonkeys-as 
I have made two different classes for `DCBTrainer` and `MABTrainer`.
Also, should I add the tutorial for contextual bandits in the same file as this or make a new one?

### CODE REVIEW

**Comment by sampreet-arthi** [comment]
Since there's a lot of different bandits, I'd suggest keeping the tutorial docs in separate files

File: docs/source/tutorials/Using MultiArmedBandits.md [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+
+trainer = MABTrainer(agent, bandit)
+trainer.train(timesteps=10000)
+```
+
+Implementation of the bandit:
+https://github.com/SforAiDl/genrl/blob/master/genrl/bandit/bandits/multi_armed_bandits/bernoulli_mab.py
+
+Implementation of the agent:
+https://github.com/SforAiDl/genrl/blob/master/genrl/bandit/agents/mab_agents/ucb.py
```
Comment (threewisemonkeys-as) [inline docs/source/tutorials/Using MultiArmedBandits.md:19]: Can you include a link to the docs instead of this. You an look at the Using A2C docs as a example.

File: genrl/bandit/trainer.py [hunk 2]
```diff
@@ -30,6 +30,106 @@ def __init__(
         self.log_mode = log_mode
         self.logger = Logger(logdir=logdir, formats=[*log_mode])
 
+    def train(self) -> None:
```
Comment (threewisemonkeys-as) [inline genrl/bandit/trainer.py:33]: For this class (`BanditTrainer`), can you inhertit `abc.ABC` to make it an abstract class. This way, you dont need to raise the NotImplementedError

**Comment by threewisemonkeys-as** [comment]
Also, you will have to edit the tests files relating to whatever changes you have made.

File: docs/source/tutorials/Using MultiArmedBandits.md (hunk 1 above)
Comment (Darshan-ko) [reply docs/source/tutorials/Using MultiArmedBandits.md:19]: The links in the Using A2C doc are giving an Error 404.
What should I put the link to?

Comment (threewisemonkeys-as) [reply docs/source/tutorials/Using MultiArmedBandits.md:19]: Woring fine for me. https://genrl.readthedocs.io/en/latest/usage/tutorials/Using%20A2C.html

Comment (threewisemonkeys-as) [reply docs/source/tutorials/Using MultiArmedBandits.md:19]: Try in incognito. It might be loading the old one?

**Comment by sampreet-arthi** [comment]
Take a look at the lgtm errors

**Comment by Darshan-ko** [comment]
I thought of just including the basic explanation of the problem and the algo. I can also add some formulae and equations if needed.

File: docs/source/tutorials/Using MultiArmedBandits.md (hunk 1 above)
Comment (Darshan-ko) [reply docs/source/tutorials/Using MultiArmedBandits.md]: The link in the doc in this repo is not working [https://github.com/SforAiDl/genrl/blob/master/docs/source/api/algorithms/genrl.deep.agents.a2c.html](https://github.com/SforAiDl/genrl/blob/master/docs/source/api/algorithms/genrl.deep.agents.a2c.html)

Comment (threewisemonkeys-as) [reply docs/source/tutorials/Using MultiArmedBandits.md]: Yeah, thats happening because its a relative link. In this, you can just link it to  the docs directly. E.g. for bernoulli MAB you can link https://genrl.readthedocs.io/en/latest/api/bandit/genrl.bandit.bandits.multi_armed_bandits.html#genrl.bandit.bandits.multi_armed_bandits.bernoulli_mab.BernoulliMAB

Comment (Darshan-ko) [reply docs/source/tutorials/Using MultiArmedBandits.md]: I've done this in the last commit.

**Comment by Darshan-ko** [comment]
What my idea was to include the basic description of the bandit problem and its setting (which was previously in the UCB doc) in the EpsGreedy doc instead, since it is the most basic agent and someone new to RL will most probably want to read that up first. People wanting to use UCB will mostly have the basic knowledge of what a bandit is and so I think it might be better in the file 'Tutorial_on_bandits' instead of 'Using_UCB'.

File: docs/source/usage/tutorials/Tutorial_on_bandits.md [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+
+We want our estimate $Q_t(a)$ to be as close to the true value $q(a)$ as possible, so we can make the correct decision.
+
+Let the action with the maximum quality be $a^*$: 
+
+$$q^* = q(a^*)$$
+
+Our goal is to find this $q^*$.
+
+The 'regret function' is defined as the sum of 'regrets' accumulated over all timesteps. This regret is the cost of not chosing the optimum arm and instead exploring. Mathematically it can be written as:
```
Comment (Sharad24) [inline docs/source/usage/tutorials/Tutorial_on_bandits.md:27]:
```suggestion
The 'regret function' is defined as the sum of 'regret' accumulated over all timesteps. This regret is the cost of not choosing the optimal arm and instead of exploring. Mathematically it can be written as:
```

File: docs/source/usage/tutorials/Tutorial_on_bandits.md [hunk 4]
```diff
@@ -0,0 +1,60 @@
+### Training an EpsilonGreedy agent on a Bernoulli Multi Armed Bandit
+
+Multi armed bandits is one of the most basic problems in RL. Think of it like this, you have 'n' levers in front of you and each of these levers has a different reward function: the probability of getting a reward when a lever is pulled. Suppose you try out one of the levers and get a positive reward. What do you do next? Do you just keep pulling that lever every time or think what if there might be a better reward to pulling one of the other levers? This is the exploration - exploitation dilemma.
```
Comment (Sharad24) [inline docs/source/usage/tutorials/Tutorial_on_bandits.md:3]:
```suggestion
Multi armed bandits is one of the most basic problems in RL. Think of it like this, you have 'n' levers in front of you and each of these levers will give you a different reward. For the purposes of formalising the problem the reward is written down in terms of a reward function i.e., the probability of getting a reward when a lever is pulled. 

Suppose you try out one of the levers and get a positive reward. What do you do next? Should you just keep pulling that lever every time or think what if there might be a better reward to pulling one of the other levers? This is the exploration - exploitation dilemma.
```

File: docs/source/usage/tutorials/Tutorial_on_bandits.md [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+
+Exploitation - You use the information you have gathered till now, to make the best decision. In this case, after 1 try you know a lever is giving you a positive reward and you just 'exploit' it further(known as the greedy action).
```
Comment (Sharad24) [inline docs/source/usage/tutorials/Tutorial_on_bandits.md:5]:
```suggestion
_Exploitation_ - Utilise the information you have gathered till now, to make the best decision. In this case, after 1 try you know a lever is giving you a positive reward and you just _exploit_ it further. Since you do not care about other arms if you keep _exploiting_, it is known as the greedy action.
```

File: docs/source/usage/tutorials/Tutorial_on_bandits.md [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+
+Exploration - You explore the untried levers in an attempt to maybe discover another one which has a higher payout than the one you currently have some knowledge about. This is exploring all your options without worrying about the short-term rewards, in hope of finding a lever with a bigger reward, in the long run.
+
+You have to use an algorithm which correctly trades off exploration and exploitation as we do not want a 'greedy' algorithm which only exploits and does not explore at all, because there are very high chances that it will converge to a sub-optimum policy. Neither do we want an algorithm which only explores, because this would lead to non-maximum rewards inspite of knowing the best action to be taken. In this case, the optimum policy will be to always pull the lever with the highest reward, but at the beginning we do not know the probability distribution of the rewards. So, we want a policy which explores actively at the beginning, building up an estimate for the reward values(defined as quality) of all the actions, and then exploiting that from that time onwards.
```
Comment (Sharad24) [inline docs/source/usage/tutorials/Tutorial_on_bandits.md:9]:
```suggestion
You have to use an algorithm which correctly trades off exploration and exploitation as we do not want a 'greedy' algorithm which only exploits and does not explore at all, because there are very high chances that it will converge to a sub-optimal policy. We do not want an algorithm that keeps exploring either as this would lead to sub-optimal rewards inspite of knowing the best action to be taken. In this case, the optimal policy will be to always pull the lever with the highest reward, but at the beginning we do not know the probability distribution of the rewards. 

So, we want a policy which explores actively at the beginning, building up an estimate for the reward values(defined as _quality_) of all the actions, and then exploiting that from that time onwards.
```

File: docs/source/usage/tutorials/Using_UCB.md [hunk 7]
```diff
@@ -0,0 +1,44 @@
+### Training a UCB algorithm on a BernoulliMulti-ArmedBandit
```
Comment (Sharad24) [inline docs/source/usage/tutorials/Using_UCB.md:1]:
```suggestion
### Training a UCB algorithm on a Bernoulli Multi-Armed Bandit
```

File: docs/source/usage/tutorials/Using_UCB.md [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+
```
Comment (Sharad24) [inline docs/source/usage/tutorials/Using_UCB.md:2]:
```suggestion
For an introduction to Multi Armed Bandits, refer to Tutorial On Bandits
```

File: docs/source/usage/tutorials/Tutorial_on_bandits.md (hunk 3 above)
Comment (Darshan-ko) [reply docs/source/usage/tutorials/Tutorial_on_bandits.md:27]: The cost is incurred due to exploring, so shouldn't it be 'intstead' and not 'instead of'? Like you do not care of the short term reward. Instead, you explore.

**Comment by Darshan-ko** [comment]
@Sharad24 @threewisemonkeys-as @sampreet-arthi 
If this is good I'll start pushing the docs I've made for other bandit algos? Any particular suggestions?

**Comment by Sharad24** [comment]
Sounds good!

File: docs/source/usage/tutorials/Tutorial_on_bandits.md [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+
+Suppose you try out one of the levers and get a positive reward. What do you do next? Should you just keep pulling that lever every time or think what if there might be a better reward to pulling one of the other levers? This is the exploration - exploitation dilemma.
+
+_Exploitation_ - Utilise the information you have gathered till now, to make the best decision. In this case, after 1 try you know a lever is giving you a positive reward and you just _exploit_ it further. Since you do not care about other arms if you keep _exploiting_, it is known as the greedy action.
+
+Exploration - You explore the untried levers in an attempt to maybe discover another one which has a higher payout than the one you currently have some knowledge about. This is exploring all your options without worrying about the short-term rewards, in hope of finding a lever with a bigger reward, in the long run.
```
Comment (threewisemonkeys-as) [inline docs/source/usage/tutorials/Tutorial_on_bandits.md:9]: this should be italicised like exploitation

File: docs/source/usage/tutorials/Tutorial_on_bandits.md [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+
+You have to use an algorithm which correctly trades off exploration and exploitation as we do not want a 'greedy' algorithm which only exploits and does not explore at all, because there are very high chances that it will converge to a sub-optimal policy. We do not want an algorithm that keeps exploring either as this would lead to sub-optimal rewards inspite of knowing the best action to be taken. In this case, the optimal policy will be to always pull the lever with the highest reward, but at the beginning we do not know the probability distribution of the rewards. 
+
+So, we want a policy which explores actively at the beginning, building up an estimate for the reward values(defined as _quality_) of all the actions, and then exploiting that from that time onwards.
+
+A Bernoulli Multi-Armed Banit has multiple arms with each having a different bernoulli distribution over its reward. Basically each arm has a probabilty associated with it which is the probability of getting a reward if that arm is pulled. Our aim is to find the arm which has the highest probabilty, thus giving us the maximum return.
```
Comment (threewisemonkeys-as) [inline docs/source/usage/tutorials/Tutorial_on_bandits.md:15]: mispell of bandit

File: docs/source/usage/tutorials/Using_UCB.md [hunk 11]
```diff
@@ -0,0 +1,44 @@
+### Training a UCB algorithm on a Bernoulli Multi-Armed Bandit
+For an introduction to Multi Armed Bandits, refer to Tutorial On Bandits
```
Comment (threewisemonkeys-as) [inline docs/source/usage/tutorials/Using_UCB.md:2]: you can link the tutorial here.

File: docs/source/usage/tutorials/Tutorial_on_bandits.md [hunk 12]
```diff
@@ -0,0 +1,64 @@
...
+arms = 5
+
+reward_probs = np.random.random(size=(bandits, arms))
+bandit = BernoulliMAB(bandits, arms, reward_probs, context_type="int")
+agent = EpsGreedyMABAgent(bandit, eps=0.05)
+
+trainer = MABTrainer(agent, bandit)
+trainer.train(timesteps=10000)
+```
+More details can be found in the docs for [BernoulliMAB](https://genrl.readthedocs.io/en/latest/api/bandit/genrl.bandit.bandits.multi_armed_bandits.html#genrl.bandit.bandits.multi_armed_bandits.bernoulli_mab.BernoulliMAB), [EpsGreedyMABAgent](https://genrl.readthedocs.io/en/latest/api/bandit/genrl.bandit.agents.mab_agents.html#module-genrl.bandit.agents.mab_agents.epsgreedy), [MABTrainer](https://genrl.readthedocs.io/en/latest/api/common/bandit.html#module-genrl.bandit.trainer).
```
Comment (threewisemonkeys-as) [inline docs/source/usage/tutorials/Tutorial_on_bandits.md:64]: You can also refer to sutton barto for further info on bandits.

**Comment by threewisemonkeys-as** [comment]
Apart from the few small suggestions, look good.

**Comment by Darshan-ko** [comment]
I did a git pull on my local copy of the branch to pull the code review changes I made here. And it doubled up the commits(all the older commits show up again). Did I do something wrong here?

**Comment by sampreet-arthi** [comment]
Is there any code that's coming twice or any other change? If not, it's probably not a problem.

**Comment by Darshan-ko** [comment]
> Is there any code that's coming twice or any other change? If not, it's probably not a problem.

No there's no repeating code, so I guess that's ok. Should I have used `git fetch` and `git rebase` instead of `git pull` to avoid this in the future?

**Comment by sampreet-arthi** [comment]
Maybe git fetch and git merge. Best would be to make changes only locally. It shouldn't be a recurring problem dw.

**Comment by Darshan-ko** [comment]
I have added the tutorials for all the agents. Maybe we can seperate the bandit agents and the other agents into two different folders under genrl/docs/source/usage/tutorials ?

**Comment by threewisemonkeys-as** [comment]
Yeah. Since there are a lot of bandit based agents, it makes sense to have a seperate section within tuorials for bandits

**Comment by threewisemonkeys-as** [comment]
Also I think you will have to edit the `tutorials/index.rst` for these files to show up in table of contents.

**Comment by threewisemonkeys-as** [comment]
@Darshan-ko, did you check if the docs look fine locally?

**Comment by Darshan-ko** [comment]
> @Darshan-ko, did you check if the docs look fine locally?

If you mean the formulae and equations, yes they look fine locally.
I'll just make a different folder for bandit tutorials and edit the index.rst file

**Comment by Darshan-ko** [comment]
@threewisemonkeys-as If I make a different folder under `tutorials` named `Bandits`, will I need to have a different index.rst file in `tutorials/Bandits` as well or just adding `Bandits` to `tutorials/index.rst` and changing the `:maxdepth: 2`will work?

**Comment by threewisemonkeys-as** [comment]
You can either  (I think, not 100% sure) -
1. reference the individual tutorials from `tutorials/index.rst` 
2. reference them in `tutorials/bandit/index.rst` and reference this file in `tutorials/index.rst` and increase depth

**Comment by Darshan-ko** [comment]
I've added an index.rst file in `tutorials/bandits` and also changed the depth in `tutorials/index.rst`. I think this will work.

File: docs/source/usage/tutorials/Bandits/index.rst [hunk 13]
```diff
@@ -0,0 +1,11 @@
...
+============
+
+.. toctree::
+    :maxdepth: 1
+
+    Tutorial_on_bandits
+    Using_UCB
+    Using_Thompson_Sampling
+    Using_Bayesian_Method
+    Using_Gradient_Method
```
Comment (Sharad24) [inline docs/source/usage/tutorials/Bandits/index.rst:11]: Do we need these to be named `Using-*`?

Comment (Darshan-ko) [reply docs/source/usage/tutorials/Bandits/index.rst:11]: Even I thought it was redundant to use `Using-*` under tutorials, but the other files were named that way. I can change that and also the heading in those files, because that's what will show up as the text on the hyperlinks on the site, I think.

Comment (sampreet-arthi) [reply docs/source/usage/tutorials/Bandits/index.rst:11]: Hmm, I guess it started with the A2C PR. We can do without the "Using" because it's in usage/tutorials anyway.

Comment (Darshan-ko) [reply docs/source/usage/tutorials/Bandits/index.rst]: @sampreet-arthi I've done it. I think everything is resolved now.

**Comment by sampreet-arthi** [comment]
@threewisemonkeys-as @Sharad24 any other changes needed?

**Comment by threewisemonkeys-as** [comment]
looks good

**Comment by sampreet-arthi** [comment]
Ok, @Sharad24 can you approve/merge

**Review by Sharad24 — APPROVED** [review]
Looks fine, but for the future we want to avoid having `try`, `except` and `finally` statements as part of the core library.

**Comment by Sharad24** [comment]
Merge at will @sampreet-arthi @threewisemonkeys-as
